File: sudoku.js
class Sudoku {
    constructor() {
        this.content = Array(81).fill(0);
        this.availables = Array(81).fill(9);
    }

    // Función toString
    toString() {
        let result = '';
        for (let i = 0; i < 9; i++) {
            result += this.content.slice(i * 9, (i + 1) * 9).join(' ') + '\n';
        }
        return result;
    }

    // Funciones básicas para añadir / borrar elementos.
    add(pos, number) {
        if (pos >= 81 || pos < 0) {
            throw new TypeError(`Invalid position: ${pos}`);
        }
        if (number <= 0 || number > 9) {
            throw new TypeError(`Invalid number: ${number}`);
        }
        

        this.content[pos] = number;
        this.updateAvailableNumbers();
    }

    remove(pos) {
        if (pos >= 81 || pos < 0) {
            throw new TypeError(`Invalid position: ${pos}`);
        }
        this.content[pos] = 0;
        this.updateAvailableNumbers();
    }

    clear() {
        this.content = Array(81).fill(0);
    }

    shuffle(array) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
    }

    isComplete() {
        return this.content.every(value => value !== 0);
    }

    isComplete_Matrix() {
        let success = true; 
        for (let i = 0; i < this.content.length; i++) {
            success = this.content[i].every(value => value !== 0);
            if (success === false) {
                return;
            }
        }
        return success;
    }
    convertToMatrix() {
        const matrix = [];
        for (let i = 0; i < this.content.length; i += 9) {
            const row = this.content.slice(i, i + 9);
            matrix.push(row);
        }
        this.content = [...matrix];
    }

    // Funciones necesarias para ubicar números y actualizar los disponibles.
    getCurrentRow(pos) {
        return Math.floor(pos / 9);
    }

    getCurrentColumn(pos) {
        return pos % 9;
    }

    getFullRow(row) {
        return this.content.slice(row * 9, row * 9 + 9);
    }

    getFullColumn(column) {
        let columnValues = [];
        for (let i = 0; i < 9; i++) {
            columnValues.push(this.content[column + 9 * i]);
        }
        return columnValues;
    }

    getCurrentSquare(pos) {
        const row = this.getCurrentRow(pos);
        const col = this.getCurrentColumn(pos);
        const square = [Math.floor(row / 3), Math.floor(col / 3)];
        return square;
    }

    getFullSquare(square) {
        const fullSquare = [];
        const startRow = square[0] * 3;
        const startCol = square[1] * 3;

        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                const pos = (startRow + i) * 9 + (startCol + j);
                fullSquare.push(this.content[pos]);
            }
        }
        return fullSquare;
    }

    getAvailableNumbers(pos) {
        const row = this.getFullRow(this.getCurrentRow(pos));
        const col = this.getFullColumn(this.getCurrentColumn(pos));
        const square = this.getFullSquare(this.getCurrentSquare(pos));

        const setRow = new Set(row);
        const setCol = new Set(col);
        const setSquare = new Set(square);

        const notAvailables = new Set([...setRow, ...setCol, ...setSquare]);
        const numbers = new Set([1, 2, 3, 4, 5, 6, 7, 8, 9]);
        const availables = new Set([...numbers].filter(num => !notAvailables.has(num)));

        return [...availables];
    }

    updateAvailableNumbers() {
        for (let i = 0; i < this.availables.length; i++) {
            if (this.content[i] === 0) {
                this.availables[i] = this.getAvailableNumbers(i).length;
            } else {
                this.availables[i] = 0;
            }
        }
    }

    getMinAvailablePos() {
        let min = 10;
        let pos = -1;

        for (let i = 0; i < this.availables.length; i++) {
            if (this.availables[i] < min && this.content[i] == 0) {
                min = this.availables[i];
                pos = i;
            }
        }

        return pos;
    }

    testUniqueness(current) {
        let firstSolution = new Sudoku();
        firstSolution.content = [...current];
        firstSolution.createRecursive(false);

        for (let i = 0; i < 10; i++) {
            let sol = new Sudoku();
            sol.content = Array.from(current);
            sol.createRecursive(false);
            if (!this.arraysEqual(sol.content, firstSolution.content)) {
                return false; // Si la encuentra, no es único.
            }
        }
        return true; // Si no encuentra una solución diferente en 10 intentos, se entiende que es único.
    }

    arraysEqual(arr1, arr2) {
        if (arr1.length !== arr2.length) return false;
        for (let i = 0; i < arr1.length; i++) {
            if (arr1[i] !== arr2[i]) return false;
        }
        return true;
    }

    createRecursive(clear=true) {
        if (clear) {
            this.clear();
        }

        let minPos = this.getMinAvailablePos();
        return this._createRecursive(minPos);
    }

    _createRecursive(pos) {
        // Caso base
        if (pos === -1 || this.content[pos] !== 0) {
            return this.content;
        }

        // Caso Recursivo
        let availableNumbers = this.getAvailableNumbers(pos);
        this.shuffle(availableNumbers);

        for (var number of availableNumbers) {
            this.add(pos, number);
            let minPos = this.getMinAvailablePos();

            if (minPos === -1 || this.isComplete()) {
                break;
            }

            this._createRecursive(minPos);

            if (this.isComplete()) {
                return this.content;
            }
            this.remove(pos);
        }
    }

    createPlayable(difficulty = 2) {
        if (![1, 2, 3].includes(difficulty)) {
            throw new Error(`Invalid difficulty: ${difficulty}. Must be in [1, 2, 3]`);
        }

        let numbersToLeave;
        if (difficulty === 3) {
            numbersToLeave = Math.floor(Math.random() * (28 - 24 + 1)) + 24; // Entre 24 y 28
        } else if (difficulty === 2) {
            numbersToLeave = Math.floor(Math.random() * (33 - 29 + 1)) + 29; // Entre 29 y 33
        } else {
            numbersToLeave = Math.floor(Math.random() * (38 - 34 + 1)) + 34; // Entre 34 y 38
        }

        const n = 81 - numbersToLeave;

        this.createRecursive();
        this.updateAvailableNumbers();
        this._createPlayable(n);

        return this.content;
    }

    _createPlayable(n, visited = []) {
        if (n <= 0 || visited.length >= 81) {
            return;
        }

        let position = Math.floor(Math.random() * 81);

        if (visited.includes(position)) {
            this._createPlayable(n, visited);
            return;
        }

        visited.push(position);

        let number = this.content[position];
        this.remove(position);

        let isUnique = this.testUniqueness(this.content);

        if (!isUnique) {
            this.add(position, number);
        } else {
            n -= 1;
        }

        this._createPlayable(n, visited);
    }
}

export default Sudoku;

// function main() {
//     let sudoku1 = new Sudoku();
//     sudoku1.createPlayable();
//     console.log(sudoku1.toString());
//     console.log(sudoku1.getFullSquare(sudoku1.getCurrentSquare(0)));
// }

// main();

File: sudoku.py
import random

class Sudoku:
    def __init__(self):
        # Crea un sudoku vacío .
        self.content = [0 for i in range(81)]
        self.availables = [9 for i in range(81)]
        self.debug = 0

     
    def __str__(self):
        # GOAT Method para imprimir los sudokus.
        result = ""
        for i in range(len(self.content)):
            if i != 0:
                # Saltos de línea
                if i % 9 == 0:
                    result += "\n"
                # Separación de cuadrados vertical
                elif i % 3 == 0 :
                    result += " │"

                # Separación de cuadrados horizontal
                if i %27 == 0:
                    result += "───────┼───────┼───────\n"  
            
            result += " "
            # Añadir los números  
            result += str(self.content[i] if self.content[i] != 0 else "·")
        result += "\n"
        return result
    
    def print_availables(self):
        # GOAT Method para imprimir los sudokus.
        result = ""
        for i in range(len(self.content)):
            if i != 0:
                # Saltos de línea
                if i % 9 == 0:
                    result += "\n"
                # Separación de cuadrados vertical
                elif i % 3 == 0 :
                    result += " │"

                # Separación de cuadrados horizontal
                if i %27 == 0:
                    result += "───────┼───────┼───────\n"  
            
            result += " "
            # Añadir los números  
            result += str(self.availables[i] if self.availables[i] != 0 else "·")
        result += "\n"
        print(result)


    ''' FUNCIONES DE AÑADIR Y BORRAR'''
    def add(self, pos, number):
        # Añade el elemento number en la posición pos.

        if pos >= 81 or pos < 0:
            raise TypeError(f"Invalid position: {pos}")
        if number <= 0 or number > 9:
            raise TypeError(f"Invalid number: {number}")
    
        self.content[pos] = number
        self.update_available_numbers()

    def remove(self, pos: int):

        #Borra el elemento que haya en una posición.
        if pos >= 81 or pos < 0:
            raise ValueError(f"Invalid position: {pos}")
        
        self.content[pos] = 0
        self.update_available_numbers()


    ''' FUNCIONES NECESARIAS PARA AVERIGUAR POSICIONES '''
    def get_current_row(self, pos: int):
        return pos // 9

    def get_full_row(self, row: int):
        # Devuelve tofos los números de la fila. Esta es la más fácil de las 3.

        return self.content[9 * row : 9 * row + 9]

    def get_current_column(self, pos: int):
        return pos % 9

    def get_full_column(self, column: int):
        # Usa list comprehension para crear la lista. 
        # Devuelve todos los números cuyas posiciones con congruentes con column (mod 9).

        return [self.content[column + 9*i ] for i in range(9)]

    def get_current_square(self, pos: int):
        current_row = self.get_current_row(pos)
        current_column = self.get_current_column(pos)
        current_square = [current_row //3, current_column // 3]
        return current_square

    def get_full_square(self, square: list[int, int]):
        # Me comí un poco la cabeza para esta.
        # Utiliza los índices necesarios para situar el cuadrado y después los añade a full_square
        # (La "primera casilla" de cada cuadrado es la SUPERIOR IZQUIERDA, para que sea todo más fácil.)

        full_square = []
        for i in range(3):
            for elem in self.content[(9 * 3 * square[0]) + 3 * square[1] + 9*i:
                                            (9 * 3* square[0]) + 3 * square[1] + 3 + 9*i] :
                full_square.append(elem)
        
        return full_square

    ''' FUNCIÓN GRANDE Y ÚTIL. '''
    def get_available_numbers(self, pos: int):
        # Encuentra la fila, columna y cuadrado de la posición pos.
        current_row = self.get_current_row(pos)
        current_column = self.get_current_column(pos)
        current_square = self.get_current_square(pos) 

        full_row =self.get_full_row(current_row)
        full_column = self.get_full_column(current_column)

        # Recopilación del cuadrado entero
        full_square = self.get_full_square(current_square)

        # Uso de sets para averiguar los elementos disponibles
        not_available = set(set(full_row) | set(full_column) |set(full_square)) #Unión de 3 conjuntos.
        available = list(set([i for i in range(1, 10)]) - not_available) 
    
        # print("Position:", current_row,"", current_column)
        # print("Square:", current_square)
        # print(not_available)
        # print("Available: ", available)

        return available


    def update_available_numbers(self):
        for i in range(len(self.content)):
            self.availables[i] = len(self.get_available_numbers(i)) if self.content[i] == 0 else 0
        # print("Availables: \n", sudoku, "\n")

    def get_min_available_number(self):
        min = 10
        for i in range(len(self.availables)):
            if  self.availables[i] < min and self.content[i] == 0: 
                min = self.availables[i]

        # Si min es 10 es porque no ha encontrado ningún número vacío.
        return min if min != 10 else -1
    
    def get_min_available_position(self):
        min = self.get_min_available_number()

        if min == -1:
            return -1
        
        # Sudoku no completo, busca la posición mínima.
        for i in range(len(self.content)):
            if self.availables[i] == min:
                return i


    def clear(self):
        #Deja el sudoku actual en blanco.
        self.content = [0 for i in range(81)]

    def check_complete(self):
        # Función que comprueba que el sudoku esté totalmente lleno de números.
        for i in self.content:
            if i == 0:
                return False
        return True     


    ''' FUNCIÓN DE CREADO'''
    def create_bruteforce(self): 
        ''' 
        Esta función se encarga de crear un sudoku completo y sin errores.
        
        El algoritmo empleado dentro del bucle while es el mejor que se me ha ocurrido
        hasta ahora, pues funciona aproximadamente el 40% de las veces.

        El bucle while lo utilizo para asegurarme de que el sudoku que he creado esté completo.

        La función NO DEVUELVE NADA, si se quiere ver el sudoku, se deberá usar print(). 
        Ej:
            sudoku1 = Sudoku()
            sudoku1.create()
            print(sudoku1) -> Sudoku completo. 

        '''
        
        complete = False
        while not complete:
            self.clear()
            self.update_available_numbers()
            for i in range(len(self.content)):
                min = self.get_min_available_number()
                # Encontrar la casilla con menor número de posibilidades.
                position = i
                for j in range(len(self.content)):
                    if self.availables[j] == min:
                        position = j
                        break
                
                avl = self.get_available_numbers(position)
                for k in range(len(avl)):
                    number = avl[random.randint(0, len(avl)-1)]
                    self.add(position, number)
                    if self.check_valid_number():
                        break

            complete = self.check_complete()   
        


    def test_uniqueness(self, current) -> bool:
        # Genera la primera solución desde el estado actual
        solution = Sudoku()
        solution.content = current.copy()  # Asegúrate de copiar el estado actual
        solution.create_recursive(clear=False)
        first_solution = solution.content.copy()
        
        # Intenta encontrar una segunda solución diferente, lo hace 10 veces.
        for i in range(10):
            solution_check = Sudoku()
            solution_check.content = current.copy()  
            solution_check.create_recursive(clear=False)
            
            # Si encontramos una solución diferente, no es única
            if solution_check.content != first_solution:
                print(self.debug)
                self.debug += 1
                return False
                
        
        # Si después de 10 intentos todas las soluciones son iguales, es única
        
        return True



    ''' FUNCIÓN DE CREAR PARA JUGAR'''
    def create_playable(self, difficulty: int = 2):
        # Borra el contenido del sudoku. Por defecto, la función lo hace.  
        # Comprobaciones iniciales.
        if difficulty not in [1, 2, 3]:
            raise ValueError(f"Invalid difficulty: {difficulty}. Must be in [1, 2, 3]")
        
        # Difficulty 3 : Deja 24- 28 números.
        # Difficulty 2: Deja 29-33 números.
        # Difficulty 1: Deja 34 - 38 números.
        numbers_to_leave = random.randint(24, 28) if difficulty == 3 else random.randint(29, 33) if difficulty == 2 else random.randint(34, 38)
        
        # n es el número de posiciones que va a borrar.
        n = 81 - numbers_to_leave
        # Se parte de un sudoku lleno.
        self.create_recursive()
        initial = self.content.copy() # Muy importante para que no guarden la misma dirección de memoria.
        self.update_available_numbers()

        # Parte recursiva
        self._create_playable(initial, n)

        return self.content

        
    def _create_playable(self, initial, n: int, positions=None):
        # Si es la primera llamada, inicializa el conjunto de posiciones posibles.
        if positions is None:
            positions = list(range(81))
            random.shuffle(positions)

        # Caso Base
        if n <= 0 or not positions:
            return 

        # CASO RECURSIVO: 
        # Coge una posición de la lista de posiciones.  
        position = positions.pop()

        # Si ya está vacía, pasa a la siguiente recursión
        if self.content[position] == 0:
            self._create_playable(initial, n, positions)
            return

        # Borra el número de la posición escogida.
        self.remove(position)

        # Comprueba que sigue teniendo solución única.
        unique = self.test_uniqueness(current=self.content)

        # Si no tiene solución única, restaura el borrado y no disminuye `n`.
        if not unique:
            self.add(position, initial[position])
        else:
            n -= 1  # Disminuye `n` solo si se borró un número válido.

        # Sigue la recursión.
        self._create_playable(initial, n, positions)


    def create_recursive(self, clear = True):
        if clear:
            self.clear()

        min_pos = self.get_min_available_position()
        
        return self._create_recursive(min_pos)

    def _create_recursive(self, pos):
        if pos == -1 or self.content[pos] != 0:
            return self.content

        available_numbers = self.get_available_numbers(pos)
        random.shuffle(available_numbers)
        for number in available_numbers:
            # Añade el número
            self.add(pos, number)

            # Busca la posición con menos números disponibles
            min_pos = self.get_min_available_position()
            if min_pos == -1 or self.check_complete():
                break
            
            # Llamada recursiva: El sudoku aún no está completo.
            self._create_recursive(min_pos)
            
            # Si después de llamar a la función detecta que está completo, para la recursión.
            if self.check_complete():
                return self.content
            
            # Si no, quita el número que había puesto.
            self.remove(pos)


    def string_to_content(self, str: str):
        # Mostly useless unless you want to create a sudoku from a previous one.
        content = []
        for character in str:
            if character.isnumeric():
                content.append(int(character))
            elif character == "·":
                content.append(0)
        if len(content) != 81:
            raise ValueError(f"Invalid length: {len(content)}.")
        

        self.content = content.copy()

    def get_all_squares(self):
        squares = [[i, j] for i in range(3) for j in range(3)]
        result = []
        for sq in squares:
            full_square = self.get_full_square(sq)
            result.append(full_square)
        return result
    
    def square_format(self):
        #Función que devuelve el contenido del sudoku en forma de matrices de cada cuadrado 3x3.
        squares = self.get_all_squares()
        result = []
        for sq in squares:
           # Convierte la lista sq en una matriz 3x3.
           result.append([sq[i: i+3] for i in range(0, 9, 3)])
        return result


if __name__ == "__main__":
    sudoku = Sudoku()
    sudoku.create_playable(difficulty=1)
    print(f"Playable sudoku: \n\n{sudoku}")

    ''' Descomentar para comprobar que el sudoku es único. '''
    # j = 0
    # sudoku_og = Sudoku()
    # sudoku_og.content = sudoku.content.copy()
    # sudoku_og.create_recursive(clear=False)
    # for i in range(100):  
    #     sudoku2 = Sudoku()
    #     sudoku2.content = sudoku.content.copy()
    #     sudoku2.create_recursive(clear=False)
    #     lista = sudoku2.content.copy()
    #     if lista != sudoku_og.content:
    #         print(sudoku_og.content,"\n\n", lista)
    #         print(f"Falso {j}")
    #         j += 1
    #print(sudoku.square_format())




    